src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2316,7 +2316,7 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatement(SyntaxKind par
                 if (!sawRef &&
                     identifierOrThisOpt != null &&
                     (typeParameterListOpt != null && typeParameterListOpt.ContainsDiagnostics
-                      || this.CurrentToken.Kind != SyntaxKind.OpenParenToken && this.CurrentToken.Kind != SyntaxKind.OpenBraceToken) &&
+                      || this.CurrentToken.Kind != SyntaxKind.OpenParenToken && this.CurrentToken.Kind != SyntaxKind.OpenBraceToken && this.CurrentToken.Kind != SyntaxKind.EqualsGreaterThanToken) &&
```

Comment:
this is getting complex enough that i think extracting woudl be valuable.  But up to you.

Reply:
One benefit of extraction is htat each case can be individually checked, with a nice comment explaining the real language construct this is looking for.  I find that that is easier to read, review, and most importantly: maintain.  Someone can see those checks and immediately start thinking to themselves "is there a another common case that is missing here?"

Reply:
For example, what if i have "async this[...]".  Do we have a problem here?  you should have "identifierOrThisOpt"... but you'd have an OpenBracketToken not an openbrace.

Reply:
Note: if the indexer case works, it woudl still probably be helpful to add tests in this area for 'async' as the return type for all sorts of language constructs.

Reply:
I should read this piece to understand why we need to `reconsider "async" to be a modifier` in the first place. With `partial` there were no such ambiguities. @CyrusNajmabadi do you have any idea what we're trying to do here?

Reply:
nope!  hence why i wish this stuff was broken out into specific, documented, cases.  Because trying to infer after the fact is super tough in these codeblocks that deal with very ambiguous code.

Reply:
Alright, turns out it was doing somersault on ambiguities that we already dealt with for `partial`. Now it's hugely simplified and gives more accurate results: https://github.com/alrz/roslyn/commit/6a4d586859559d7a949aef8b64b518f4593f4e0b

I figured the `ScanPartial` acts as a `IsTrueModifier` - just gives more details. that's why we can share many parts with `ShouldAsyncTreatedAsModifier`, however, the later has some specific preference that make these two not so identical. I'll open a PR for further discussion when other two are merged.